_posts/work/2018-07-08-dask-dev.md

```diff
@@ -107,25 +107,28 @@ incremental or batchwise fashion.
 ```python
 from sklearn.linear_model import SGDClassifier
 
-sgd = SGDClassifier(...)
+est = SGDClassifier(...)
 
 import pandas as pd
 
-for filename in filenames:
-    df = pd.read_csv(filename)
-    X, y = ...
+dfs = [pd.read_csv(filename) for filename in filenames]
 
-    sgd.partial_fit(X, y)
+for X, y in fit_data(dfs):
+    est.partial_fit(X, y)
 ```
 
+This requires reading in the entire the entire dataset on one machine,
+and keeps the dataset and the model on one machine. It explicitly moves
+each dataset to the model.
```

Comment:
If I were to read the entire data into memory, as in the `dfs = [...]` line then wouldn't I just call fit?

Reply:
Ah, I see where you're coming from. Let me explain where I'm coming from.

For `SGDClassifier`, it's important that `partial_fit` calls have a random sample from the *entire* dataset. Otherwise, it's possible that the weights are not the same between one `fit` call and many `partial_fit` calls.

For example, take the case where every CSV records different behavior and one `partial_fit` calls converges to that CSV's solution: then many `Incremental.partial_fit` calls hop around between the local minima for each CSV, and never gets to the global minima for all CSVs. I  believe this case would be likely, especially for CSVs.

I've updated and try to explain this, and point to https://github.com/dask/dask/issues/3409.